cmd/wshd/main.go

```diff
@@ -33,15 +36,116 @@ const (
 	topicAnswererOnCandidate = "wampshell.webrtc.answerer.on_candidate"
 )
 
+type interactiveShellSession struct {
+	ptmx map[uint64]*os.File
+	sync.Mutex
+}
+
+func newInteractiveShellSession() *interactiveShellSession {
+	return &interactiveShellSession{
+		ptmx: make(map[uint64]*os.File),
+	}
+}
+
+func (p *interactiveShellSession) handleShell(e *wampshell.EncryptionManager) func(_ context.Context,
+	inv *xconn.Invocation) *xconn.InvocationResult {
+	return func(_ context.Context, inv *xconn.Invocation) *xconn.InvocationResult {
+		caller := inv.Caller()
+
+		e.Lock()
+		key, ok := e.Keys()[inv.Caller()]
+		e.Unlock()
+		if !ok {
+			return xconn.NewInvocationError("wamp.error.unavailable", "unavailable")
+		}
+
+		p.Lock()
+		ptmx, ok := p.ptmx[caller]
+		p.Unlock()
+
+		if !ok {
+			cmd := exec.Command("bash")
+			newPtmx, err := pty.Start(cmd)
+			if err != nil {
+				return xconn.NewInvocationError("io.xconn.error", err.Error())
+			}
+
+			ptmx = newPtmx
+			p.Lock()
+			p.ptmx[caller] = newPtmx
+			p.Unlock()
+
+			go func(inv *xconn.Invocation, ptmx *os.File) {
+				buf := make([]byte, 4096)
+				for {
+					n, err := ptmx.Read(buf)
+					if n > 0 {
+						ciphertext, nonce, errEnc := berncrypt.EncryptChaCha20Poly1305(buf[:n], key.Send)
+						if errEnc != nil {
+							log.Printf("Encryption failed in shell output: %v", errEnc)
+							break
+						}
+						payload := append(nonce, ciphertext...)
+						_ = inv.SendProgress([]any{payload}, nil)
+					}
+
+					if err != nil {
+						_ = inv.SendProgress(nil, nil)
+						break
+					}
+				}
+
+			}(inv, ptmx)
+			return xconn.NewInvocationError(xconn.ErrNoResult)
+
```

Comment:
we can create a separate function for this may be called `startPty` or `startTerminal`